### DISCUSSION TITLE
BiasFHV metrics calculation

### DISCUSSION THREAD

**Discussion by SebaDro** [neuralhydrology/neuralhydrology#94] [Q&A] [answered]
Hi all,

during some literature research I had a closer look on some of the metrics implementations. The implementation of the BiasFHV metric is according to your [documentation ](https://neuralhydrology.readthedocs.io/en/latest/api/neuralhydrology.evaluation.metrics.html#neuralhydrology.evaluation.metrics.fdc_fhv):

$\text{BiasFHV} = \frac{\sum_{h=1}(Q_{s,h} - Q_{o,h})}{\sum_{h=1}Q_{o,h}} \times 100$

This correctly relates to the formulation in [Yilmaz et al. (2008)](https://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/2007WR006716)

However, in your [implementation](https://github.com/neuralhydrology/neuralhydrology/blob/a4c284b92934ee973c8b3fedf8a60df60c8feae1/neuralhydrology/evaluation/metrics.py#L419-L469) I was wondering about the calculation of the indices of the top 2 % flows. Since the FDC is calculated separately for observations and simulations, I was wondering whether this procedure is correct. It is likely to happen that the extracted values for observations and simulations relates to different timesteps in the timeseries.

Another approach would be to only determine the indices of the top 2 % values of the observed flow and use these indices to extract the values of the simulated flow. My understanding of the BiasFHV metric is that it expresses of how good is a model to predict the 2 % highest observed values. This means the model predictions for the same 2 % value timesteps should be used for this purpose. 

[Yilmaz et al. (2008)](https://agupubs.onlinelibrary.wiley.com/doi/abs/10.1029/2007WR006716) says:
> (...) where h = 1, 2,. . .H are the flow indices for flows with exceedance probabilities lower than 0.02.

This also may be interpreted differently. What do you think about it?

Thanks.

**Comment by danklotz** [comment, marked as answer]
Hi [SebaDro](https://github.com/SebaDro), 

your observations (that the observations are not compared to corresponding simulations) often comes up in our discussions and I believe that I can say that we all share your concerns. However, I feat that this a property of the definition of the $BiasFHV$ and not an implementation detail. I provide the following reasoning for this: (1) Yilmaz, Gupta and Wagener (2008) base their metrics on the FDC (see l37ff) and it would be peculiar if the $BiasFHV$ is the only metric that uses a different indexing. (2) The start of the sentence that you quote (l60 in the appendix) reads: 


> The diagnostic signature measure for long-term baseflow is the percent bias in FDC low-segment volume (%BiasFLV) ...


which, in my eyes tells us that two parts of the FDC are compared. 

That said, I am happy to be convinced from the opposite if you have additional arguments. 

---

**To summarize:** I (and all of our team) share your critique, but I fear that the $BiasFHV$ is faithfully implemented. However, it might still make sense to implement another high-flow metrics that works similar to the one you just described (say a high-flow bias HFB). I am looking forward to get a pull request that does so :)

**Reply by SebaDro** [reply]
Hi @danklotz,

first of all, many thanks for your response. 

Actually, I can follow your reasoning. And I have to admit, that if you base all the signature measures on the FDC, you only can come to the conclusion that the metrics calculation is based on the comparison of certain segments of two different FDCs (the observed FDC and the simulated FDC). This leads to the fact that observed flows are not compared to the corresponding simulated flows for the same timesteps. 

I think, the technical derivation of all the signature measures is the problem here. The authors take the overlapping of observed FDC and simulated FDC as a goodness measure for model performance. However, the shortcome of this approach is that it does not take into account the temporal order of observed and simulated flows. This only would be correct, if you assume a linear correlation between observed and simulated flows (a high observed flow value at one timestep has a similiar high simulated flow value).
In the worst case, two FDCs may be nearly identical although simulated flows completely differs from the corresponding observations.

Summarizing, I share your and your team's opinion that the $\text{BiasFHV}$ (and all other signature measures that are described in Yilmaz, Gupta and Wagener (2008)) are faithfully implemented in your library. But at the same time, I doubt that these measures solely provide a meaningful performance measure in some situations. Hence, these measures always should be contextualized with regression-based metrics (RMSE, NSE, KGE, ...), which you also provide in your library.

In addition, I agree that new metrics that compare corresponding high and low flows would be very helpful. So, I'd be happy if I could contribute to your framework with such an implementation.

**Reply by kratzert** [reply]
Hi, great seeing some discussion here. I just wanted to say that I share all of what was said here. Only two things I want to add:

- There exist also a peak-timing metric in NeuralHydrology https://github.com/neuralhydrology/neuralhydrology/blob/master/neuralhydrology/evaluation/metrics.py#L538
- I would still very much welcome a PR that adds a peak-bias metric (e.g. mean peak bias in percent). If you have time to add that, that would be amazing and I would be happy to help you with the entire PR process. Otherwise I might implement something in this direction myself soon, as I most likely need something like that.

**Reply by SebaDro** [reply]
Hey,

thanks for sharing your thoughts about this. Yes, I'd really like to come up with such peak-bias metric. You can expect a first draft in a few days 😄.
